### ISSUE TITLE
Crash exiting fullscreen (F11) in Linux with Wayland and Nvidia drivers

### ISSUE THREAD

**Issue by JPLeBreton** [Robbendebiene/Gesturefy#679]
### Describe your matter in detail
Steps to reproduce:
1. Launch latest version of addon in latest Firefox on a GNOME + Wayland + Nvidia system.
2. On any webpage, press F11 to enter fullscreen
3. Press F11 again to exit fullscreen
4. Observe: Firefox crashes

It seems highly likely this is an Nvidia bug, a Firefox bug, or something else rather than a Gesturefy bug, but I'm filing it here first to narrow down the issue. When I log in to an X11 session rather than a Wayland session, the crash doesn't happen. When I load a fresh Firefox profile without Gesturefy installed, the crash doesn't happen. When I install addons other than Gesturefy on this new profile, the crash still doesn't happen.

Happy to provide the firefox crash log or any other info needed.

### Your System
- Operating system: Manjaro Linux (stable branch) x86-64
- Firefox version: 116.0.2
- Gesturefy version: 3.2.7
- Video drivers + GPU: Nvidia binary 535.98 on a GTX 1070

**Comment by Robbendebiene** [comment]
Hi, thanks for your detailed report.
I'm very surprised that Gesturefy can cause such a problem. The only relation I can currently think of is that Gesturefy uses the [Canvas API](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API ) to draw the gesture trace. Obviously this could be hardware accelerated and ultimately resulting in a crash.

However if this is the case other websites that make use of the canvas API should crash the browser too.

Just as an idea, can you try to disable the gesture trace in the Gesturefy settings and see if the crash still occurs? (Gesturefy will still create the canvas element but it won't append it to the DOM)

Maybe there is some relation to this: https://www.reddit.com/r/firefox/comments/11cnto2/psa_linux_users_may_need_to_disable_accelerated/

**Comment by JPLeBreton** [comment]
> However if this is the case other websites that make use of the canvas API should crash the browser too.
> Just as an idea, can you try to disable the gesture trace in the Gesturefy settings and see if the crash still occurs? (Gesturefy will still create the canvas element but it won't append it to the DOM)

Aha, this is exactly the kind of info I was hoping for.
Disabling the gesture trace and restarting Firefox makes the crashes go away.
Leaving gesture trace on and setting gfx.canvas.accelerated=false *also* makes the crashes go away.

I tried some web-based Canvas demos (with gfx.canvas.accelerated=true again), but going in and out of fullscreen with them running didn't cause any crashes. That makes me wonder how different the code paths are for browser extensions drawing to a canvas vs web pages (which might be more sandboxed or scrutinized by the browser in various ways? i don't know much about this stuff).

**Comment by JPLeBreton** [comment]
I spoke too soon! Leaving this demo https://webglsamples.org/aquarium/aquarium.html running in a tab for a while, doing other stuff, then going in and out of fullscreen crashes firefox. So that's a repro that (I'm reasonably sure) doesn't involve this extension's functionality. So it's either a Firefox bug or something further down the stack. Thanks very much for the troubleshooting help, this was enough to give me a general direction. Feel free to close this issue out.

**Comment by Robbendebiene** [comment]
Appreciate all the information and understanding.

Out of interest, does Firefox immediately crash when you visit a website with Gesturey's gesture trace enabled or only when you start "drawing" a gesture?

Does [drawing here](https://leimao.github.io/blog/HTML-Canvas-Mouse-Touch-Drawing/) crash your Firefox as well?

In case you are going to open a bugzilla ticket or found an existing one, please post a comment with the link here :)

**Comment by JPLeBreton** [comment]
Okay, here's the bugzilla I created: https://bugzilla.mozilla.org/show_bug.cgi?id=1850285

If this ultimately ends up being an Nvidia issue and leads me to their support forum, I'll post that here as well.

**Comment by Robbendebiene** [comment]
Thank you. So for now Gesturefy somehow provides the most reliable way to reproduce this bug?

I just quickly extracted the parts where I thought that they may cause the crash. Maybe you can execute the below code in a website of your choice (pasting it into the tab's console that can be open via ctrl + shift + I and then switch to Fullscreen). But maybe there is more to it than the below code. What makes me a little confident is that Gesturefy has some logic that is only triggered on resize events (it resizes the canvas and restores the canvas properties).

```js
// use HTML namespace so proper HTML elements will be created even in foreign doctypes/namespaces (issue #565)

const Overlay = document.createElementNS("http://www.w3.org/1999/xhtml", "div");
      Overlay.style = `
        all: initial !important;
        position: fixed !important;
        top: 0 !important;
        bottom: 0 !important;
        left: 0 !important;
        right: 0 !important;
        z-index: 2147483647 !important;

        pointer-events: none !important;
      `;

const Canvas = document.createElementNS("http://www.w3.org/1999/xhtml", "canvas");
      Canvas.style = `
        all: initial !important;

        pointer-events: none !important;
      `;

const Context = Canvas.getContext('2d');

// resize canvas on window resize
window.addEventListener('resize', maximizeCanvas, true);
maximizeCanvas();

initialize();

function initialize () {
  Overlay.append(Canvas);

  // overlay is not working in a pure svg or other xml pages thus do not append the overlay
  if (!document.body && document.documentElement.namespaceURI !== "http://www.w3.org/1999/xhtml") {
    return;
  }
  // if an element is in fullscreen mode and this element is not the document root (html element)
  // append the overlay to this element (issue #148)
  if (document.fullscreenElement && document.fullscreenElement !== document.documentElement) {
    document.fullscreenElement.appendChild(Overlay);
  }
  else if (document.body.tagName.toUpperCase() === "FRAMESET") {
    document.documentElement.appendChild(Overlay);
  }
  else {
    document.body.appendChild(Overlay);
  }
}


function maximizeCanvas () {
  // save context properties, because they get cleared on canvas resize
  const tmpContext = {
    lineCap: Context.lineCap,
    lineJoin: Context.lineJoin,
    fillStyle: Context.fillStyle,
    strokeStyle: Context.strokeStyle,
    lineWidth: Context.lineWidth
  };

  Canvas.width = window.innerWidth;
  Canvas.height = window.innerHeight;

  // restore previous context properties
  Object.assign(Context, tmpContext);
}
```

**Comment by JPLeBreton** [comment]
> Thank you. So for now Gesturefy somehow provides the most reliable way to reproduce this bug?

The simplest repro I've found so far is just: launch with a clean profile (no Gesturefy installed), open the HTML canvas aquarium demo in a tab, then *let it run for a while* (unclear how long), then maaybe do other stuff outside the browser on my computer, then try F11-ing back in firefox? I'm trying to narrow down the simplest possible repro steps, and am going to try to eliminate everything except the canvas tab + time passing by leaving a fresh session running overnight.

I ran that code you posted in the browser console, and exiting fullscreen right after didn't immediately trigger it. But I'll keep it in mind for future repro attempts.
